connect-location/src/main/java/com/ifttt/location/GeofenceProvider.java

```diff
@@ -0,0 +1,115 @@
+
+                    for (String fenceKey: fenceKeys) {
+                        if (!updatedIds.contains(fenceKey)) {
+                            requestBuilder.removeFence(fenceKey);
+                        }
+                    }
+
+                    for (LocationTriggerField locationTriggerField: locationTriggerFields) {
+                        LocationFieldValue region = locationTriggerField.locationFieldValue;
+
+                        PendingIntent exitPendingIntent = PendingIntent.getBroadcast(
+                                context,
+                                REQUEST_CODE_EXIT,
+                                new Intent(context, AwarenessExitReceiver.class),
+                                PendingIntent.FLAG_UPDATE_CURRENT
+                        );
+
+                        PendingIntent enterPendingIntent = PendingIntent.getBroadcast(
+                                context,
+                                REQUEST_CODE_ENTER,
+                                new Intent(context, AwarenessEnterReceiver.class),
+                                PendingIntent.FLAG_UPDATE_CURRENT
+                        );
+
+                        switch (locationTriggerField.permissionName) {
+                            case ConnectLocation.TRIGGER_ID_LOCATION_ENTER:
+                                requestBuilder.addFence(
+                                        locationTriggerField.id,
+                                        LocationFence.entering(region.lat, region.lng, region.radius),
+                                        enterPendingIntent
+                                );
+                                break;
+
+                            case ConnectLocation.TRIGGER_ID_LOCATION_EXIT:
+                                requestBuilder.addFence(
+                                        locationTriggerField.id,
+                                        LocationFence.exiting(region.lat, region.lng, region.radius),
+                                        exitPendingIntent
+                                );
+                                break;
+
+                            case ConnectLocation.TRIGGER_ID_LOCATION_ENTER_EXIT:
+                                requestBuilder.addFence(
+                                        getExitFenceKey(locationTriggerField.id),
+                                        LocationFence.exiting(region.lat, region.lng, region.radius),
+                                        exitPendingIntent
+                                );
+                                requestBuilder.addFence(
+                                        getEnterFenceKey(locationTriggerField.id),
+                                        LocationFence.entering(region.lat, region.lng, region.radius),
+                                        enterPendingIntent
+                                );
+                        }
+                        fenceClient.updateFences(requestBuilder.build());
```

Comment:
This might be something we should try: instead of always re-registering the geofences, we should only do so if them are not already. The documentation does mention that we only need to re-register in certain scenarios.

Reply:
We only need to re-register for reboot scenario. The logic to not add already registered fences to the request builder should be fairly straightforward, however I want to test this version end to end before we add that.